code/datums/abilities/flock/flockmind.dm

```diff
@@ -286,55 +359,116 @@
-	var/structurewantedtype = null
 	var/turf/T = get_turf(holder.owner)
 	if(!istype(T, /turf/simulated/floor/feather))
 		boutput(holder.owner, "<span class='alert'>You aren't above a flocktile.</span>")//todo maybe make this flock themed?
-		return 1
+		return TRUE
+	if(locate(/obj/flock_structure/ghost) in T)
+		boutput(holder.owner, "<span class='alert'>A tealprint has already been scheduled here!</span>")
+		return TRUE
 	if(locate(/obj/flock_structure) in T)
 		boutput(holder.owner, "<span class='alert'>There is already a flock structure on this flocktile!</span>")
-		return 1
+		return TRUE
+
+	var/list/friendlyNames = list()
+	var/mob/living/intangible/flock/flockmind/F = holder.owner
+	for(var/datum/unlockable_flock_structure/ufs as anything in F.flock.unlockableStructures)
+		if(ufs.check_unlocked())
+			friendlyNames += ufs.friendly_name
+
+
 	//todo: replace with FANCY tgui/chui window with WHEELS and ICONS and stuff!
-	var/structurewanted = tgui_input_list(holder.owner, "Select which structure you would like to create", "Tealprint selection", list("Collector", "Sentinel"))
-	switch(structurewanted)
-		if("Collector")
-			structurewantedtype = /obj/flock_structure/collector
-			resourcecost = 200
-		if("Sentinel")
-			structurewantedtype = /obj/flock_structure/sentinel
-			resourcecost = 300
+
+	var/structurewanted = tgui_input_list(holder.owner, "Select which structure you would like to create", "Tealprint selection", friendlyNames)
+
+	if (!structurewanted)
+		return TRUE
+	var/obj/flock_structure/structurewantedtype = null
+	for(var/datum/unlockable_flock_structure/ufs as anything in F.flock.unlockableStructures)
+		if(ufs.friendly_name == structurewanted)
+			structurewantedtype = ufs.structType
+			break
+
 	if(structurewantedtype)
-		var/mob/living/intangible/flock/F = holder.owner
-		F.createstructure(structurewantedtype, resourcecost)
+		return F.createstructure(structurewantedtype, initial(structurewantedtype.resourcecost))
+
+/////////////////////////////////////////
+
+/datum/targetable/flockmindAbility/ping
+	name = "Ping"
+	desc = "Request attention from other elements of the flock."
+	icon_state = "ping"
+	cooldown = 0.1 SECONDS
+
+/datum/targetable/flockmindAbility/ping/cast(atom/target)
+	if(..())
+		return TRUE
+	if (!isturf(target.loc) && !isturf(target))
+		return TRUE
+	var/mob/living/intangible/flock/F = holder.owner
+	F.flock?.ping(target, holder.owner)
+
+/////////////////////////////////////////
+
+/datum/targetable/flockmindAbility/deconstruct
+	name = "Mark for Deconstruction"
+	desc = "Mark an existing flock structure for deconstruction, refunding some resources."
+	icon_state = "destroystructure"
+	cooldown = 0.1 SECONDS
+
+/datum/targetable/flockmindAbility/deconstruct/cast(atom/target)
+	if(..())
+		return TRUE
+	var/mob/living/intangible/flock/F = holder.owner
+	if(HAS_ATOM_PROPERTY(target,PROP_ATOM_FLOCK_THING))
+		if (isflockdeconimmune(target)) // ghost structure on click opens tgui window
```

Comment:
how is the comment relevant?
(I'm sure there's some reason but I'm not seeing it)

Reply:
For ghost structures they already have a Click() that open a TGUI prompt for deconstruction, this prevents toggling deconstruction for ghost structures (which also affects flockdrone AI)

Reply:
Ah, makes sense